tests/std/tests/P0896R4_ranges_alg_transform_binary/test.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+
+#include <range_algorithm_support.hpp>
+
+using ranges::transform, ranges::binary_transform_result, ranges::iterator_t;
+using std::identity, std::same_as;
+using P = std::pair<int, int>;
+
+// Validate that binary_transform_result aliases in_in_out_result
+STATIC_ASSERT(same_as<binary_transform_result<int, int, double>, ranges::in_in_out_result<int, int, double>>);
+
+// Validate dangling story
+STATIC_ASSERT(
+    same_as<decltype(transform(borrowed<false>{}, borrowed<false>{}, static_cast<int*>(nullptr), std::plus{})),
+        binary_transform_result<ranges::dangling, ranges::dangling, int*>>);
+STATIC_ASSERT(same_as<decltype(transform(borrowed<false>{}, borrowed<true>{}, static_cast<int*>(nullptr), std::plus{})),
+    binary_transform_result<ranges::dangling, int*, int*>>);
+STATIC_ASSERT(same_as<decltype(transform(borrowed<true>{}, borrowed<false>{}, static_cast<int*>(nullptr), std::plus{})),
+    binary_transform_result<int*, ranges::dangling, int*>>);
+STATIC_ASSERT(same_as<decltype(transform(borrowed<true>{}, borrowed<true>{}, static_cast<int*>(nullptr), std::plus{})),
+    binary_transform_result<int*, int*, int*>>);
+
+constexpr auto minus_one = [](int x) -> int { return x - 1; };
+constexpr auto cast      = [](double x) -> int { return static_cast<int>(x); };
+
+struct binary_transform_test {
+    static constexpr std::array<P, 3> input1      = {{{1, 99}, {4, 98}, {5, 97}}};
+    static constexpr std::array<double, 3> input2 = {{6.4, 7.9, 8.2}};
+    static constexpr std::array<int, 3> expected  = {{7, 11, 13}};
+
+    template <ranges::input_range In1, ranges::input_range In2, class Write>
+    static constexpr void call() {
+        { // Validate range overload
+            std::array<int, 3> output{};
+            In1 wrapped_in1{input1};
+            In2 wrapped_in2{input2};
+
+            auto result = transform(wrapped_in1, wrapped_in2, Write{output.data()}, std::plus<int>{}, get_first, cast);
+            STATIC_ASSERT(same_as<decltype(result), binary_transform_result<iterator_t<In1>, iterator_t<In2>, Write>>);
+            assert(result.in1 == wrapped_in1.end());
+            assert(result.in2 == wrapped_in2.end());
+            // assert(result.out == Write{output.data() + 3});

```

Comment:
It appears that the chosen `writable_iterator`'s are not comparable by default. Thoughts on what to do here?

I thought about using `std::addressof` but that gives a type error between `const int*` and `int*`

@CaseyCarter: It is a bit unfortunate that the iterator constructor only takes pointers and not iterators. Could we change that?

Reply:
Specializations of `test::iterator` have a `base()` member that provides access to the wrapped pointer.